### PR TITLE
Check your answers

### PR DESCRIPTION
**Pull request by michaeldfallen**
This PR adds a new CheckYourAnswers step class that will allow the user to summarise and check what they've entered before submission.

**CheckYourAnswers step**

To create a check your answers step:

- extend CheckYourAnswers and hook it in to your journey as normal

    ```
    const { CheckYourAnswers } = require('@hmcts/one-per-page/checkYourAnswers');

    class CheckAppeal extends CYA {}
    ```

- create a template that extends the check your answers base template

    ```
    {% extends "look-and-feel/layouts/check_your_answers.html" %}

    {% block head -%}
      <link href="{{ asset_path }}main.css" media="screen" rel="stylesheet" />
    {% endblock %}

    {% set phase="ALPHA" %}
    {% set feedbackLink="https://github.com/hmcts/one-per-page/issues/new" %}
    ```

- Hook in the check your answers stylesheets

    _main.scss_
    ```
    @import 'look-and-feel/check-your-answers';
    ```

**Declaring answers**

By default your Questions will output their class Name and their fields concatenated on check your answers.

If you want to customise the output you can override the answers a Question produces like so:

```
class Country extends Question {
  ...
  answers() {
    return answer(this, {
      // capitalise first letter of country
      answer: titleise(this.fields.country.value)
    });
  }
  ...
}
```

The answer function takes the following options:

- **`answer`** the actual answer output on check your answers
- **`question`** the content to render for the question, e.g. You could put `Your name` instead of just `Name`
- **`values`** currently unused, will be used for online submission
- **`section`** the section to put the answer in. Will be placed in the default section if left out.
- **`url`** the url to send the user to if they want to change the answer, defaults to the steps path
- **`hide`** whether to hide the answer from rendering on check your answers

A step can generate multiple answers, simply return an array from the answers function:

```
  answers() {
    return [
      answer(this, { question: 'Your name', ... }),
      answer(this, { question: 'Partners name', ... })
    ];
  }
```

**Sectioning Check Your Answers**

Complex journeys will want to split the answers in to logical blocks. 

- Declare the sections in CheckYourAnswers

```
const { section } = require('@hmcts/one-per-page/checkYourAnswers');

class CheckAppeal extends CheckYourAnswers {
  sections() {
    return [
      section('personal-details', { title: 'Personal Details' }),
      section('respondent-details', { title: 'Respondent Details' })
    ];
  }
}
```

- In each step declare what section the answer should be placed in

_Name.step.js_
```
class Name extends Question {
  ...
  answers() {
    return answer(this, { section: 'personal-details' });
  }
  ...
}
```

- Any answers you don't declare a section on will render in the "Other details" section

### CODE REVIEW

File: test/views/question_views/simpleQuestion.html [hunk 1]
```diff
@@ -9,7 +9,7 @@ <h1>Question 1</h1>
   {% endif %}
 
   {{ fields.name }}
-  <input name="{{ fields.name.id }}"
+  <input name="{{ fields.name.name }}"
```
Comment (endamccormack) [inline test/views/question_views/simpleQuestion.html]: value?

File: src/flow/walkTree.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+const walkTree = (start, steps) => {
+  let current = start;
+  let iterations = 0;
+  const results = [];
+  const numberOfSteps = Object.keys(steps).length;
+  while (iterations <= numberOfSteps) {
+    if (current instanceof Question) {
+      results.push(current);
+      current.retrieve().validate();
```
[the remark this answers is not in the document]
Comment (michaeldfallen) [reply src/flow/walkTree.js]: Yep, it's an assumption that the validation tells us whether a step is complete. If a step isn't complete then we stop walking the tree and use that step as the URL to "Continue your application". 

That might not be a safe assumption but it seems to work ATM.

File: src/forms/field.js [hunk 3]
```diff
@@ -93,9 +76,13 @@ class FieldDesriptor {
    *   session
    */
   serialize() {
-    if (typeof this.id === 'undefined') return {};
-    if (typeof this.value === 'undefined') return {};
-    return { [this.id]: this.value };
+    if (notDefined(this.value)) {
```
[the remark this answers is not in the document]
Comment (michaeldfallen) [reply src/forms/field.js]: Good call. I've been meaning to pull other common checks in to the checks file as well (`isString`, `isObject`, etc)

File: examples/test-app/steps/Country.step.js [hunk 4]
```diff
@@ -1,18 +1,35 @@
 const { Question, goTo, branch } = require('@hmcts/one-per-page');
 const { form, textField } = require('@hmcts/one-per-page/forms');
+const { answer } = require('@hmcts/one-per-page/checkYourAnswers');
 const Joi = require('joi');
 
-class Country extends Question {
-  get url() {
-    return '/country';
+
+const titleise = string => {
```
[the remark this answers is not in the document]
Comment (michaeldfallen) [reply examples/test-app/steps/Country.step.js]: yeah, I wanted to keep the example app separate from the library. so it only imports from one-per-pages declared interfaces.

File: src/forms/field.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (michaeldfallen) [reply src/forms/field.js]: I'll do this in a separate PR, 43 files changed to implement it.

File: src/middleware/resolveTemplate.js [hunk 5]
```diff
@@ -31,38 +31,43 @@ const stepTargets = step => {
...
-  return [
-    'No templates found',
-    '',
-    ...attemptLines
-  ].join('\n');
+  return new Error(
+    [
+      'No templates found',
+      '',
+      ...attemptLines
```
[the remark this answers is not in the document]
Comment (michaeldfallen) [reply src/middleware/resolveTemplate.js]: An error that is thrown if no templates are found. It outputs all the paths it tried so the developer can better understand how the framework thinks. The `...` specifically is the es6 spread syntax

File: test/views/question_views/simpleQuestion.html (hunk 1 above)
Comment (michaeldfallen) [reply test/views/question_views/simpleQuestion.html]: The name of the field that the input should be posted to on a POST.

**Pull request merged**